lib/ansible/modules/windows/win_dhcp_lease.ps1

```diff
@@ -0,0 +1,473 @@
+        # Insert Dashes
+        $mac = $mac.Insert(2, "-").Insert(5, "-").Insert(8, "-").Insert(11, "-").Insert(14, "-")
+        return $mac
+    }
+    elseif ($mac.Length -eq 17) {
+        # Remove Dashes
+        $mac = $mac -replace '-'
+        return $mac
+    }
+    else {
+        return $false
+    }
+}
+
+Function Compare-DhcpLease {
+    Param(
+        [PSObject]$Original,
+        [PSObject]$Updated
+    )
+
+    # Compare values that we care about
+    if (($Original.AddressState -eq $Updated.AddressState) -and ($Original.IPAddress -eq $Updated.IPAddress) -and ($Original.ScopeId -eq $Updated.ScopeId) -and ($Original.Name -eq $Updated.Name) -and ($Original.Description -eq $Updated.Description)) {
+        # changed = false
+        return $false
+    }
+    else {
+        # changed = true
+        return $true
+    }
+}
+
+Function Convert-ReturnValue {
+    Param(
+        $Object
+    )
+
+    $data = @{
+        address_state = $Object.AddressState
+        client_id     = $Object.ClientId
+        ip_address    = $Object.IPAddress.IPAddressToString
+        scope_id      = $Object.ScopeId.IPAddressToString
+        name          = $Object.Name
+        description   = $Object.Description
+    }
+
+    return $data
+}
+
+# Parse Regtype
+if ($dns_regtype) {
+    Switch ($dns_regtype) {
+        "aptr" { $dns_regtype = "AandPTR"; break }
+        "a" { $dns_regtype = "A"; break }
+        "noreg" { $dns_regtype = "NoRegistration"; break }
+        default { $dns_regtype = "NoRegistration"; break }
+    }
+}
+
+Try {
+    # Import DHCP Server PS Module
+    Import-Module DhcpServer
+}
+Catch {
+    # Couldn't load the DhcpServer Module
+    $module.FailJson("The DhcpServer module failed to load properly")
+}
+
+# Determine if there is an existing lease
+if ($ip) {
+    $current_lease = Get-DhcpServerv4Scope | Get-DhcpServerv4Lease | Where-Object IPAddress -eq $ip
+}
+
+# MacAddress was specified
+if ($mac) {
+    if ($mac -like "*-*") {
+        $mac_original = $mac
+        $mac = Convert-MacAddress -mac $mac
+    }
+
+    if ($mac -eq $false) {
+        $module.FailJson("The MAC Address is not properly formatted")
+    }
+    else {
+        $current_lease = Get-DhcpServerv4Scope | Get-DhcpServerv4Lease | Where-Object ClientId -eq $mac_original
+    }
+}
+
+# Did we find a lease/reservation
+if ($current_lease) {
+    $current_lease_exists = $true
+    $original_lease = $current_lease
+    $module.Diff.before = Convert-ReturnValue -Object $original_lease
+}
+else {
+    $current_lease_exists = $false
+}
+
+# If we found a lease, is it a reservation?
+if ($current_lease_exists -eq $true -and ($current_lease.AddressState -like "*Reservation*")) {
+    $current_lease_reservation = $true
+}
+else {
+    $current_lease_reservation = $false
+}
+
+# State: Absent
+# Ensure the DHCP Lease/Reservation is not present
+if ($state -eq "absent") {
+
+    # Required: MAC or IP address
+    if ((-not $mac) -and (-not $ip)) {
```

Comment:
This can be implemented by adding `required_if = @(,@("state", "absent", @("mac", "ip")))` to the module spec. The format for a `required_if` entry is

```
@(
    @("option to check",
        "value of option",
        @("list of other options where one must be set if the option is set to value"),
    )
)
```